designs/node-overlay.md

```diff
@@ -0,0 +1,355 @@
+  requirements:
+  - key: karpenter.k8s.aws/instance-memory
+    operator: Gt
+    values: [2048]
+  capacity:
+    memory: 50Mi
+---
+kind: NodeOverlay
+metadata:
+  name: extended-resource
+spec:
+  # weight is not required since default does not conflict with extended-resource
+  requirements:
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ["m5.large", "m5.2xlarge", "m5.4xlarge", "m5.8xlarge", "m5.12xlarge"]
+  capacity:
+    smarter-devices/fuse: 1
+```
+
+### Misconfiguration
+
+When multiple NodeOverlays have conflicting configurations, particularly with identical weights, we must choose between two fundamental approaches: failing open or failing closed for Karpenter’s scale and disruption actions. Each approach presents different tradeoffs for system reliability and operator experience.
+
+**Fail Open Approach (recommended):**
+
+* Using alphabetical ordering or validate last applied resource to resolve conflicts when weights are equal
+* Setting overlay Ready status condition to false when conflicts occur
+* Adding status conditions to inform users that the overlays aren't applied to any instance types
+* Continuing provisioning/consolidation operations even when overlay configurations don't match users intent
+
+While this approach maintains system availability. Failing open could lead to unexpected resource allocations and pricing decisions, potentially impacting workload performance and cost optimization goals. Furthermore, if an unapplied overlay defines extended resources (like HugePages), pods requiring these resources would fail to get provisioned capacity - matching Karpenter's current behavior when required resources are unavailable. This means that while the system remains operational, workloads depending on specific overlay-defined resources would still be blocked until the configuration issues are resolved.
+
+```
+kind: NodeOverlay
+metadata:
+  name: memory-1
+spec:
+  weight: 90
+  requirements:
+  - key: karpenter.k8s.aws/instance-memory
+    operator: Gt
+    values: ["m5.large", "m5.2xlarge"]
+  capacity:
+    memory: 50Mi
+status:
+  condition:
+  - lastTransitionTime: "..."
+    message: "conflict with overlay capacity-1"
+    reason: "conflict"
+    status: "False"
+    type: Ready
+---
+kind: NodeOverlay
+metadata:
+  name: capacity-1
+spec:
+  weight: 90
+  requirements:
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ["c5.large", "c5.2xlarge", "m5.2xlarge"]
+  capacity:
+    memory: 100Mi
+status:
+  condition:
+  - lastTransitionTime: "..."
+    message: ""
+    reason: ...
+    status: "True"
+    type: Ready
+```
+
+**Fail Closed Approach:**
+
+* Halts new node launches until misconfigurations are resolved
+* Can be implemented at two levels:
+
+    1. Global: Stops all provisioning across the cluster
+    2. NodePool-specific: Only affects misconfigured NodePools
+
+* Similar to Karpenter's existing behavior with .spec.limits and InsufficientCapacityErrors
```

Comment:
How is this comparable to ICE errors?

Reply:
Similar to hitting capacity limits, this ICEd instance types today precise control over provisioning by allowing Karpenter to halt operations either across entire nodepools or for specific instance types affecting all nodepools. The solution provides flexible management of Karpenter's scheduling behavior, ensuring that resource allocation strictly adheres to intended configurations while preventing unplanned provisioning outside defined parameters.